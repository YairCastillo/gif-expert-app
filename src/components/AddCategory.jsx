import { useState } from "react";
import PropTypes from 'prop-types';

export const AddCategory = ( {onNewCategory} ) => {
  const [inputValue, setInputValue] = useState('');

  // OnChange maneja un evento, ese evento se pasa a esta funcion, y desestructuramos y extraemos solo el target
  const onInputChange = ({target}) => {
    setInputValue(target.value);
  }

  const onSubmit = (evento) => {
    // console.log(evento);
    evento.preventDefault();

    if(inputValue.trim().length <= 1) return;

    // Una de las ventajas que ofrece el useState con esta funcion es poder mandar un callback donde, en este caso, podemos tener las categorias anteriores:
    // setCategories(categories => [...categories, inputValue]);

    onNewCategory(inputValue.trim());
    setInputValue('');
  }

  return (
    <form onSubmit={ onSubmit } aria-label="form">
      <input
        type="text"
        placeholder="Buscar gifs"
        value={inputValue}
        onChange={ onInputChange }
      />
    </form>
  )
}

AddCategory.propTypes = {
  onNewCategory: PropTypes.func.isRequired
}
