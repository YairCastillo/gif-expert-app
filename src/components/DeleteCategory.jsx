import PropTypes from 'prop-types'

export const DeleteCategory = ({ onDelete, category }) => {
  return (
    <>
      <button onClick={ () => onDelete(category) } >Eliminar</button>
    </>
  )
}

DeleteCategory.propTypes = {
  onDelete: PropTypes.func.isRequired,
  category: PropTypes.string.isRequired,
}
