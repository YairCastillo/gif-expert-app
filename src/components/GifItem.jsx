export const GifItem = ({url, title}) => {
  return (
    <div className='card'>
      <img src={url} alt={title} />
      { title != ' ' && <p>{title}</p>}
    </div>
  )
}
