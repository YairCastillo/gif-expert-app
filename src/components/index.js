export * from './AddCategory';
export * from './GifGrid';
export * from './GifItem';

//Esto es un archivo de barril, cuando se haga referencia a esta carpeta (components) se busca automaticamente el index
