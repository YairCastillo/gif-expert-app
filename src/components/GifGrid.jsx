// import { useState, useEffect } from "react";
import { GifItem } from "./GifItem";
// import { getGifs } from "../helpers/getGifs";
import { useFetchGifs } from "../hooks/useFetchGifs";
import { DeleteCategory } from "./DeleteCategory";

export const GifGrid = ({ category, onDelete }) => {
  // Movido a cutom hook useFetchGifs:
  // const [images, setImages] = useState([]);

  // const getImages = async() => {
  //   const newImages = await getGifs(category);
  //   setImages(newImages);
  // }

  // Esto (la ejecucion de una funcion) NO debe hacerse dentro de un functional component porque vuelve ejecutarla (en este caso vuelve a hacer la peticion http) cada vez que cambia el estado y se vuelve a renderizar
  // getGifs(category);

  // useEffect es un hook de React que sirve para disparar efectos secundarios (un proceso que se ejecuta cuando algo suceda). useEffect NO puede devolver otra cosa que no sea una funcion. Para el estado de las imagenes (en este caso el getGifs devuelve una promesa) podriamos usar then u otra funcion
  // Movido a cutom hook useFetchGifs:
  // useEffect( () => {
  //   getImages();
  // }, []);
  // Dejar el arreglo de dependecias vacio significa que solo se ejecutara cuando se crea por primera vez el componente

  // Para todo lo anterior se puede utilizar un custom hook
  const { images, isLoading } = useFetchGifs( category );
  // console.log(isLoading);

  return (
    <>
      <div className="categoria-encabezado-container">
        <h3>{ category }</h3>
        <DeleteCategory onDelete={ onDelete } category={category} />
      </div>

      {
        // if corto con una sola condicion (AND logico)
        isLoading && (<h4>Cargando...</h4>)
      }
      

      {/* className es el sinonimo de class si fuera HTML. Como es JS, "class" es una palabra reservada */}
      <div className="card-grid">
        {
          images.map((image) => (
            <GifItem
              key={image.id}

              // Con esto, todas las propiedades que tenga el image, se esparcen en el GifItem, ideal para cuando se tienen/necesitan muchas propiedades
              { ...image }
            />
          ))
        }
      </div>
    </>
  )
}
