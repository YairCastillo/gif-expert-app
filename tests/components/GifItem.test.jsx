import { GifItem } from "../../src/components/GifItem";
const { render, screen } = require("@testing-library/react");

describe('Pruebas en GifItem', () => {

  const title = 'HAPPY DAY!!!';
  const url = 'https://www.google.com/';

  test('Debe de hacer match con el snapshot y existir el title y url', () => {
    const { container } = render( <GifItem title={title} url={url} /> );
    expect(container).toMatchSnapshot();

    expect( screen.getByText(title) ).toBeTruthy();

    const image = document.querySelector('img');
    expect(image.src).toContain(url);
  });

  test('Debe mostrar la imagen con el URL y el ALT indicado', () => {
    render( <GifItem title={title} url={url} /> );
    //screen.debug();

    // Podriamos hacer esto:
    // expect(screen.getByRole('img').src).toBe(url);
    // expect(screen.getByRole('img').alt).toBe(title);
    // o mejor:
    const {alt, src} = screen.getByRole('img');
    expect(alt).toBe(title);
    expect(src).toBe(url);
  });

  test('Debe mostrar el titulo en el componente', () => {
    render( <GifItem title={title} url={url} /> );
    expect(screen.getByText(title)).toBeTruthy();
  });
})
