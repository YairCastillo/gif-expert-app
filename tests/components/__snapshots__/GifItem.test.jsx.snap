// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en GifItem Debe de hacer match con el snapshot y existir el title y url 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="HAPPY DAY!!!"
      src="https://www.google.com/"
    />
    <p>
      HAPPY DAY!!!
    </p>
  </div>
</div>
`;
